30-indoor-air-quality-and-gas-leak/firmware/sparrow-application/aqi/aqi.cpp

```diff
@@ -0,0 +1,496 @@
+
+
+
+// We have no viable way of detecting whether or not the PIR sensor
+// hardware is present, so we use the presence of the BME280 as a proxy.
+bool isSparrowReferenceSensorBoard (void) {
+    bool result;
+
+    // Power on the sensor to see if it's here
+    GPIO_InitTypeDef init;
+    memset(&init, 0, sizeof(init));
+    init.Speed = GPIO_SPEED_FREQ_LOW;
+    init.Pin = BME_POWER_Pin;
+    init.Mode = GPIO_MODE_OUTPUT_PP;
+    init.Pull = GPIO_NOPULL;
+    HAL_GPIO_Init(BME_POWER_GPIO_Port, &init);
+    HAL_GPIO_WritePin(BME_POWER_GPIO_Port, BME_POWER_Pin, GPIO_PIN_SET);
+    MX_I2C2_Init();
+    result = (MY_I2C2_Ping(BME280_I2C_ADDR_PRIM, BME280_I2C_TIMEOUT_MS, BME280_I2C_RETRY_COUNT)
+           || MY_I2C2_Ping(BME280_I2C_ADDR_SEC, BME280_I2C_TIMEOUT_MS, BME280_I2C_RETRY_COUNT));
+    MX_I2C2_DeInit();
+
+    return result;
+}
+
+// Scheduled App One-Time Init
+bool aqiInit()
+{
+    // Register the app
+    schedAppConfig config = {
+        .name = "aqi",
+        .activationPeriodSecs = 60 * 60,    // 1 hour
+        .pollPeriodSecs = 15,
+        .activateFn = NULL,
+        .interruptFn = NULL,
+        .pollFn = aqiPoll,
+        .responseFn = aqiResponse,
+        .appContext = &aqiSensor,
+    };
+    if (schedRegisterApp(&config) < 0) {
+        return false;
+    }
+
+    // Success
+    return true;
+
+}
+
+bool isAlert(AQISensorReading& reading) {
+    return (ALERT_AQI_LEVEL && reading.aqi >= ALERT_AQI_LEVEL)
+    || (ALERT_ECO2_LEVEL && reading.eco2 >= ALERT_ECO2_LEVEL)
+    || (ALERT_TVOC_LEVEL && reading.tvoc >= ALERT_TVOC_LEVEL);
+}
+
+bool takeReading(AQISensorReading& reading) {
+    if (!aqiSensor.read(reading)) {
+        return false;
+    }
+
+    reading.alert = isAlert(reading);
+    return true;
+}
+
+bool shouldSendMonitoringEvent() {
+    // this could be used to send monitoring events less frequently than how often the app takes a sensor reading.
+    return !ALERTS_ONLY;
+}
+
+
+// Poller
+void aqiPoll(int appID, int state, void *appContext)
+{
+    // Unused parameter(s)
+    (void)appContext;
+
+    AQISensorReading sensorReading;
+
+    // Switch based upon state
+    switch (state) {
+
+    case STATE_ACTIVATED:
+        if (!templateRegistered) {
+            registerNotefileTemplate();
+            schedSetCompletionState(appID, STATE_AQI_MONITORING, STATE_ACTIVATED);
+            APP_PRINTF("aqi: template registration request\r\n");
+            break;
+        }
+        // fallthrough to monitoring immediately
+
+    case STATE_AQI_MONITORING:
+         if (takeReading(sensorReading)) {
+            if (sensorReading.isAlert()) {
+                addNote(sensorReading, true, ALERT_FIRST);
+                schedSetCompletionState(appID, STATE_AQI_ALERTING, STATE_AQI_MONITORING);
+            }
+            else if (shouldSendMonitoringEvent()) {
+                addNote(sensorReading, false, ALERT_NONE);
+                schedSetCompletionState(appID, STATE_AQI_MONITORING, STATE_AQI_MONITORING);
```

Comment:
It seems like we never deactivate this Sparrow app, from my read of the state machine. That's not necessarily a problem, but I wonder if it's working against the Sparrow architecture? Should we purposefully deactivate and then reactivate periodically (using the activation period) to take another reading?

Reply:
I think you only need to deactivate it if you're waiting for an external event so that it doesn't need to poll. But here, we don't have any interrupts so it stays active to monitor at the given periodicity. If the polling period is say 1 minute then the device will sleep in a WFI() until there is something to do.

Reply:
Fair enough.